prow/tide/tide.go

```diff
@@ -573,6 +584,86 @@ func filterPR(ghc githubClient, sp *subpool, pr *PullRequest) bool {
 	return false
 }
 
+// mergeChecker provides a function to check if a PR can be merged with
+// the requested method and does not have a merge conflict.
+// It caches results and should be cleared periodically with clearCache()
+type mergeChecker struct {
+	config config.Getter
+	ghc    githubClient
+
+	sync.Mutex
+	cache map[config.OrgRepo]map[github.PullRequestMergeType]bool
+}
+
+func newMergeChecker(cfg config.Getter, ghc githubClient) *mergeChecker {
+	m := &mergeChecker{
+		config: cfg,
+		ghc:    ghc,
+		cache:  map[config.OrgRepo]map[github.PullRequestMergeType]bool{},
+	}
+
+	go m.clearCache()
+	return m
+}
+
+func (m *mergeChecker) clearCache() {
+	// Only do this once per token reset since it could be a bit expensive for
+	// Tide instances that handle hundreds of repos.
+	ticker := time.NewTicker(time.Hour)
+	for {
+		<-ticker.C
+		m.Lock()
+		m.cache = make(map[config.OrgRepo]map[github.PullRequestMergeType]bool)
+		m.Unlock()
+	}
+}
+
+func (m *mergeChecker) repoMethods(orgRepo config.OrgRepo) (map[github.PullRequestMergeType]bool, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	repoMethods, ok := m.cache[orgRepo]
+	if !ok {
+		fullRepo, err := m.ghc.GetRepo(orgRepo.Org, orgRepo.Repo)
+		if err != nil {
+			return nil, err
+		}
+		repoMethods = map[github.PullRequestMergeType]bool{
+			github.MergeMerge:  fullRepo.AllowMergeCommit,
+			github.MergeSquash: fullRepo.AllowSquashMerge,
+			github.MergeRebase: fullRepo.AllowRebaseMerge,
+		}
+		m.cache[orgRepo] = repoMethods
+	}
+	return repoMethods, nil
+}
+
+// isAllowed checks if a PR does not have merge conflicts and requests an
+// allowed merge method. If there is no error it returns a string explanation if
+// not allowed or "" if allowed.
+func (m *mergeChecker) isAllowed(pr *PullRequest) (string, error) {
```

Comment:
Why can't we simply react to the failed merge? AKA if a merge fails, then paste a comment on the PR about why (and blacklist the PR until it is updated).

This strategy has transparency (and therefore educational) advantages. I suspect the design will be simpler and consume fewer tokens.

Reply:
Tide uses status contexts to convey information about the state of the PR. We could use a comment instead for this case, but I think it makes more sense to use the status so we don't add a comment that contradicts the status context. We also don't want to waste time retesting a PR that won't be mergeable.
We don't have a mechanism for blacklisting problematic PRs, there is a separate issue for that: https://github.com/kubernetes/test-infra/issues/10001

It is unclear to me how that strategy has more transparency than this. I think displaying a status that immediately indicates the PR has an invalid merge method is more transparent than waiting until Tide tries to merge the PR. We would definitely consume fewer tokens listing allowed merge methods, but we would waste time and potentially other tokens by considering and retesting PRs that won't be able to merge.

Reply:
What about https://github.com/kubernetes/test-infra/issues/10001#issuecomment-455638750 if we hit that error?

Just posting `/hold` means we do not really need to add a blacklisting mechanism, we just re-use what we already have.  The comment could then also include a the info that the `/hold` should be removed if the PRs or repos merge method was updated.

Reply:
Thats a good idea, but I think there are still too many downsides. We would still waste time retesting an unmergeable PR and most importantly we could get into an untested state if that PR was part of a batch, but we aren't able to merge it and accidentally merged a partial batch.

This also assumes that hook is deployed and the hold plugin is enabled. We could give Tide its own version of the hold label as Steve suggested, but then Tide would need to scrape comments or handle webhooks in order to expose a way for users to remove the label.

It seems cleanest to me to make Tide aware of this merge requirement just like the other merge requirements so that it can display it consistently with the others on the status context and properly handle the PR instead of retesting hopelessly and adding logic to recover from preventable failures.